### PR TITLE
Exercise 3

### PR DESCRIPTION
**Pull request by TimWooSoftwire**

### CODE REVIEW

File: .env.test [hunk 1]
```diff
@@ -0,0 +1,5 @@
+TRELLO_KEY = "dummy_key"
+TRELLO_TOKEN = "dummy_token"
+FLASK_SKIP_DOTENV = True
+TRELLO_BOARD_ID=5ef48df22d51c825656d5f8c
```
Comment (Jonesey13) [inline .env.test:4]: Do we really need this magic id?

File: README.md [hunk 2]
```diff
@@ -32,4 +32,16 @@ Now visit [`http://localhost:5000/`](http://localhost:5000/) in your web browser
 
 ## Secrets
 
-Secrets (a trello API key and Token) are stored in a secrets.py file, and not checked in to git.  
+Secrets (a trello API key and Token) are stored in a secrets.py file, and not checked in to git.  These look like 
+```bash
+TRELLO_KEY = "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
+TRELLO_TOKEN = "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
+```
```
Comment (Jonesey13) [inline README.md:39]: You'll also need to mention the trello ids in `config.py`

File: README.md [hunk 3]
```diff
@@ -32,4 +32,16 @@ Now visit [`http://localhost:5000/`](http://localhost:5000/) in your web browser
...
+
+## Testing
+
+Unit, integration and end-to-end tests using pylint are available - run then in your favourite IDE. 
+End to end tests will require:
+ * internet connectivity
+ * `geckodriver.exe` to be placed at the root of this repo
```
Comment (Jonesey13) [inline README.md:46]: This is currently committed into source control (so should be removed)

File: app.py [hunk 4]
```diff
@@ -1,28 +1,53 @@
...
+
+import constants
 from trello_helper import TrelloAPI as trello
+from view_model import ViewModel
+
+
+def create_app():
+    app = Flask(__name__)
+
+    constants.BOARD_ID = os.environ.get('TRELLO_BOARD_ID')
```
Comment (Jonesey13) [inline app.py:13]: Shouldn't this be reading the board id from the `config.py` file? (This is also not documented in the README.md)

File: app.py [hunk 5]
```diff
@@ -1,28 +1,53 @@
...
+    dont_check_for_list_ids = os.environ.get('DONT_CHECK_FOR_TRELLO_LIST_IDS')
```
Comment (Jonesey13) [inline app.py:14]: Usage of this env variable is undocumented

File: tests/geckodriver.log [hunk 6]
```diff
@@ -0,0 +1,84 @@
+1595596696361	mozrunner::runner	INFO	Running command: "C:\\Program Files\\Mozilla Firefox\\firefox.exe" "-marionette" "-foreground" "-no-remote" "-profile" "C:\\Users\\TimWoo\\AppData\\Local\\Temp\\rust_mozprofilezeSt59"
```
Comment (Jonesey13) [inline tests/geckodriver.log:1]: This file should be in the .gitignore (along with the binary)

File: tests/test_e2e.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+import app
+import constants as c
+from threading import Thread
+
+from selenium import webdriver
+
+
+@pytest.fixture(scope="module")
+def driver():
+    with webdriver.Firefox(r"C:\Users\TimWoo\Work\Training\DevOpsApprenticeship\ProjectExercise\DevOps-Course-Starter") as driver:
```
Comment (Jonesey13) [inline tests/test_e2e.py:13]: You can just use `Firefox()` as described in the project exercise brief

File: tests/test_e2e.py [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+    # Move to done
+    move_to_done_button = driver.find_element_by_css_selector("input.btn-success")
+    move_to_done_button.click()
+
+    # Delete
+    delete_todo_button = driver.find_element_by_css_selector("input.btn-danger")
+    delete_todo_button.click()
+
+    # The existence of the above buttons is enough to show that the card is successfully being created and moving
```
Comment (Jonesey13) [inline tests/test_e2e.py:71]: Needs some assertions (you should probably check the page state at the end)

File: tests/test_integration.py [hunk 9]
```diff
@@ -0,0 +1,45 @@
+import requests
+from dotenv import load_dotenv, find_dotenv
+import pytest
+import app
+
```
Comment (Jonesey13) [inline tests/test_integration.py:5]: Needed to add a `__init__.py` for this to run at all without import errors

File: tests/test_integration.py [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+    with test_app.test_client() as client:
+        yield client
+
+
+class MockResponse:
+    # mock json() method always returns a specific testing dictionary
+    @staticmethod
+    def json():
+        x = [
+                {"id": "1", "idList": "5ef48eae04ceb72b492830ef", "name": "ItemInTodo", "dateLastActivity": "2020-07-24T08:05:34.790Z"},
```
Comment (Jonesey13) [inline tests/test_integration.py:24]: idLists here should refer back to the constants file.

**Comment by TimWooSoftwire** [comment]
Markups are here! Apologies for delays

File: trello_helper.py [hunk 11]
```diff
@@ -25,7 +25,21 @@ def delete_item(id):
...
+        requests.put(f"{c.BASE_API_URL}/cards/{id}", params=request_payload) 
+        return id
+
+    @staticmethod
+    def get_list_ids(board_id):
+        request_payload = c.DEFAULT_PAYLOAD
+        r = requests.get(f"{c.BASE_API_URL}/boards/{board_id}/lists", params=request_payload)
+        a = r.json()
+        x = [list['id'] for list in r.json()]
+        return x
```
Comment (hugh-emerson) [inline trello_helper.py:45]: These variable names aren't the best. And the last one is pointless you can just return from the line before.

**Review by hugh-emerson — APPROVED** [review]
Looks good. I've left some small comments but go ahead and merge when you're ready.

**Pull request merged**